docs/refman/ppuser.8.pod
#
# mouse:~ppr/src/docs/refman/ppuser.8.pod
# Copyright 1995--2003, Trinity College Computing Center.
# Written by David Chappell.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
# * Redistributions of source code must retain the above copyright notice,
# this list of conditions and the following disclaimer.
# 
# * Redistributions in binary form must reproduce the above copyright
# notice, this list of conditions and the following disclaimer in the
# documentation and/or other materials provided with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE 
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
# POSSIBILITY OF SUCH DAMAGE.
#
# Last modified 15 January 2003.
#

=head1 NAME

ppuser - PPR user database management utility

=head1 SYNOPSIS

B<ppuser> [I<subcommand>] [I<parameters ... >]

=head1 DESCRIPTION

This program is used to manage PPR user charge accounts.  Ordinary users may
run the B<ppuser show> subcommand.  All other subcommands require that the
user be "root", "ppr", or be listed in F</etc/ppr/acl/ppuser.allow>.

=over 4

=item B<ppuser add> 'I<user name>' 'I<real name>' I<authcode> I<initialbalance> I<cutoff> I<life>

Adds a user to the accounting database.  "I<User name>" is the name of the
user as it will appear in "C<%%For:>" comments and L<ppr(1)> B<-f> switches.
"I<Real name>" is the name "B<user name>" should be translated to before it
is used in queue listings or banner pages.  (For example, if "I<user name>"
is "jandrews", "I<real name>" might be "Joseph Andrews. When B<ppr> is
invoked by a unix user, this field is ignored in favour of the
F</etc/passwd> gecos field, but this field will eventually be used for jobs
received from personal computers.) I<Authcode> is the password which will
identify the user. IInitialbalance> is the user's starting balance. 
I<Cutoff> is the balance below which printing privledges should be denied. 
I<Life> is the lifetime of the account.  It is measured in days of
inactivity.  Accounts are not deleted automatically, a separate program must
be employed to compare the period of inactivity to the life parameter.  Such
a program is not included in the current distribution.

=item B<ppuser delete> 'I<user name>'

delete the indicated user.

=item B<ppuser show> 'I<user name>'

show the indicated user's account information.

=item B<ppuser authcode> 'I<user name>' I<authcode>

set the indicated users authorization code (password) to \f2authcode\f1.

=item B<ppuser deposit> 'I<user name>' I<amount>

deposit money in amount I<amount> in the indicated user's account.

=item B<ppuser withdraw> 'I<user name>' I<amount>

withdraw the indicated amount I<amount> of money from the indicated
account.  This command should be used when refunding unused money.

=item B<ppuser charge> 'I<user name>' I<amount>

assess a charge of I<amount> against the indicated account.  PPR does
this automatically when a job is printed on a printer for which charges are
made.

=item B<ppuser correction> 'I<user name>' I<amount>

make a correction to the indicated user's account.  If I<amount> is
positive, the amount of money in the user's account will increase, if it is
negative, the amount of money in the account will decrease.

=back

=head1 EXAMPLES

To set up a new account for John Jones with an initial deposit of $10.00, a
cutoff when he is overdrawn by more than $2.50 and to have his account
disappear if it is unused for a year:

C<$ /usr/lib/ppr/bin/ppuser add "John Jones" "John Jones" 10.00 -2.50 365>

Or, if Mr. Jones will be entering an account-style name instead of his real
name:

C<$ /usr/lib/ppr/bin/ppuser add jjones "John Jones" 10.00 -2.50 365>

To set his password to "secret":

C<$ /usr/lib/ppr/bin/ppuser authcode "John Jones" secret>

=head1 FILES

The user account database is kept in the file "F</etc/ppr/charge_users.db>".

If the file "F</var/spool/ppr/logs/ppuser>" exists and is writable by the
user "ppr", then all B<ppuser> commands are logged there.

=head1 DIAGNOSTICS

Exit codes for ppuser are defined in the source code file
"F<include/util_exits.h>".

=head1 BUGS

If you choose to compile PPR without GDBM, this program will be
non-functional.

=head1 SEE ALSO

L<ppr(1)>, L<pprd(8)>, "PPR, a PostScript Print Spooler", "Installing and Using PPR".

=head1 HISTORY

PPR was written at Trinity College during 1993--2003.
It was first released to the public on 26 April 1995.

=head1 AUTHOR

David Chappell, Trinity College Computing Center, Hartford, Connecticut.

=cut


History.txt
mouse:~ppr/src/History.txt (symbolic link)

This file contains a history of changes to PPR which occured after the
release of version 1.00.


1.00, 26 April 1995

First public release.


1.10, 27 April 1995--7 June 1995

Removed "From:" line from message passed to Sendmail by pprd when sending
a message listing the remaining printer faults.  This is so that Sendmail
will take the "real name" from the password file and put it in the "From:"
line.

Removed dotmatrix/input.epson, a GIF file which had been inadvertently
included in the distribution.

Moved the contents of include/atalk_streams.h into
interfaces/atalk_streams.c since it was not long enough to justify an
include file, nor was it of interest to any other program.

Added code to ppad to classify members of the group "ppradmin" as PPR
addministrators who can edit printer and group configurations.

Added -v switch to ppad.

Removed uid's 1-99 from the list of priveledged users.

Fixed ppad(8) so that it considers the group "ppradmin" as priveledged
rather than "ppr".

Changed ppr(1) so that it will accept the members of the group "pprprox" as
priveledged as well as the members of "ppr".  This is for security reasons
as one may not want to grant a user who has the privedge of using the -f
switch the priveledge of deleting other people's jobs.

Modified ppop(1) and pprd(8) so that user's can request to see the queue
entry for a single job.  Previously, the only options were all jobs for a
certain destination or all jobs in all queues.

Rewrote and renamed the undocumented "ppop qdebug" command to become the new
documented command "ppop details".  (This was done primarily to support
wwwppop, an experimental managment interface for PPR.  Since it was still
incomplete it was not released with this version of PPR.)

Renamed interfaces/atalk_streams.c to atalk_ali.c and
papsrv/papsrv_streams.c to papsrv_ali.c

Removed code from "make dirs" so that it doesn't create the directory
"/etc/ppr/groups.ppd".

Added sleep=N and eoftimeout=N options to tcpip interface.

Added -C _filename_ switch to ppr invokation in xppr.

Added "-default okay" to xmessage invokation in xppr.

Added the "ppop lpq" sub-command to the "ppop help" information, thus
making it a documented sub-command.

Extended automatic bin selection to multiple media types when "%%PageMedia:"
comments are present.  This has not yet been well tested.

Made "%%EndProcset" an acceptable substitute for "%%EndProcSet" due to a
reputed bug in QuickDraw GX.

Changed rule in ppr_infile.c for detecting PCL code so that documents with
only one escape would still be detected.


1.20, 8 June 1995--10 January 1996

Fixed code in ppr_infile.c so that the ppr -o switch will work even if there
is no "DefFiltOpts:" line in the printer or group configuration file.

Added '(' to list of likely second characters for HP escape codes
in ppr_infile.c

Replaced printing of errno value with strerror(errno) output in some places,
eliminated printing of errno=x from other places where both were printed.

Fixed ppr(1) so that it sets the environment variables "PATH", "IFS", and
"SHELL" so that filters and responders it invokes are less easy to subvert.

Removed setting of PATH from ~ppr/src/misc_filters/tex.perl and changed
shell name from "/usr/bin/tainperl" to "/usr/bin/perl".

Changed code which determins default responder address so that the
environment variables "USER" and "LOGNAME" are used in preference to the
current uid.

Added prompts in Configure to set the directory into which the links to PPR
binaries will be written and the path to the X-Windows binaries such as
xterm.

Modified papsrv so that it will pass the network and node number of the
client to ppr with the -r switch.  (This is for use with an experimental
AppleTalk responder.)

Modified papsrv.c, renaming cleanup() to termination_handler() and causing
it to call fatal() only the first time it is called.  Without this, name
could be left on the network if the CAP version received two terminate
signals in rapid sucession.

Added code to re-line-wrap responder messages for certain responders.
(The library function responder_width() returns the width for a responder.
Currently it is hard coded to return zero for all responders except "samba".
A return value of zero indicates that the message should not be
re-line-wrapped.)

Dropt reading of LPDEST and PRINTER by ppr(1).  (I can not really remember
why I put this feature in.  I don't think the reasons where very good.  It
may have been that I thought that the same printer would be available thru
two spooling systems so one might want to set the default destination for
both with one variable.  That now seems a little far fetched.)

Separated Solaris and Starserver sections of sysdep.h

Added PNG (Portable Network Graphics) format with magic number recognition
to ppr_infile.c.  This remains untested since I do not have any programs
which manipulate this graphics format.

In pprd.c, put sleep(1) between execl("pprdrv") and _exit(242) so that pprd
has time to record the PID before the child exits.  (If this is not done
and execl() fails immediately, a child which should have become pprdrv may
exit before the parent records its PID in the information structure for that
queue.  This result is that it will not recoginize the purpose of the child
when it receives notification of its death and will leave the queue in the
printing state indefinitely.)

Did same in pprdrv.c in code which executes interface.  Also changed
_exit(255) to _exit(242).  (Exit 242 is the new standard since PERL uses 255
as the exit code when die() is executed.)

Added automatic creation of "DefFiltOpts:" lines to ppad(8).  Manual
creation is supported by the the "ppad deffiltopts" and "ppad group
deffiltopts" command.

Modified ppop so that interactive mode restores the terminal settings when
it exits.  Also changed terminal setting code to use the POSIX method of

ppuser/ppuser.c
/*
** mouse:~ppr/src/ppuser/ppuser.c
** Copyright 1995--2004, Trinity College Computing Center.
** Written by David Chappell.
**
** Redistribution and use in source and binary forms, with or without
** modification, are permitted provided that the following conditions are met:
** 
** * Redistributions of source code must retain the above copyright notice,
** this list of conditions and the following disclaimer.
** 
** * Redistributions in binary form must reproduce the above copyright
** notice, this list of conditions and the following disclaimer in the
** documentation and/or other materials provided with the distribution.
** 
** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
** AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
** IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
** ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE 
** LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
** CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
** SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
** INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
** CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
** ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
** POSSIBILITY OF SUCH DAMAGE.
**
** Last modified 24 May 2004.
*/

/*
** This is the operator's user account manipulation program.
*/

#include "before_system.h"
#include <sys/stat.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>
#include <string.h>
#include <stdarg.h>
#include <pwd.h>
#include <errno.h>
#include <fcntl.h>
#include <ctype.h>
#ifdef INTERNATIONAL
#include <locale.h>
#include <libintl.h>
#endif
#include "gu.h"
#include "global_defines.h"
#include "userdb.h"
#include "util_exits.h"

/* File where we log transactions: */
#define PPUSER_LOGFILE LOGDIR"/ppuser"

/*
** Handle non-fatal errors.
** This is called from functions in libpprdb.
*/
void error(const char *message, ... )
	{
	va_list va;

	va_start(va, message);
	vfprintf(stderr, message, va);
	fprintf(stderr, "\n");
	va_end(va);
	} /* end of error() */

/*
** Function for logging ppuser commands.  The log will be a valid
** shell script playing back the commands.
**
** This code was contributed by Klaus vom Scheidt
** <vscheidt@sun10rz2.rz.uni-leipzig.de>.
**
** called for:	ppuser delete ...
**				ppuser deposit ...
**				ppuser withdraw ..
**				ppuser charge ...
**				ppuser correction ...
**
** The log is written to the file LOGFILE"/ppuser",
** if that file exists.
*/
static void log_ppuser(const char *fmt, ...)
	{
	va_list va;
	FILE *file_ppuser;
	int file_ppuser_fd;

	if((file_ppuser_fd = open(PPUSER_LOGFILE, O_WRONLY | O_APPEND)) >= 0)
		{
		if((file_ppuser = fdopen(file_ppuser_fd, "a")) == NULL)
			{
			fprintf(stderr, "log_ppuser(): fdopen() failed");
			close(file_ppuser_fd);
			}
		else
			{
			va_start(va, fmt);
			fprintf(file_ppuser, "# %s \n", datestamp() );
			vfprintf(file_ppuser, fmt, va);
			va_end(va);
			fclose(file_ppuser);
			}
		}
	} /* end of log_ppuser() */

/*
** Return true if the user running this program is one that is
** allowed to perform privledged operations such as making
** deposits and posting charges.
*/
static gu_boolean privledged(void)
	{
	static int answer = -1;		/* -1 means undetermined */

	if(answer == -1)			/* if undetermined */
		{
		uid_t uid = getuid();
		answer = 0;				/* Start with "false". */

		/* Of course, "ppr" is privledged, as is uid 0 (root). */
		if(uid == 0 || uid == geteuid())
			{
			answer = 1;			/* Change answer to "true". */
			}

		/* As are all users who are members of the group "ppr". */
		else
			{
			struct passwd *pw;

			if((pw = getpwuid(uid)) == (struct passwd *)NULL)
				{
				fprintf(stderr, "privledged(): getpwuid() failed, errno=%d (%s)\n", errno, gu_strerror(errno));
				exit(EXIT_INTERNAL);
				}

			if(user_acl_allows(pw->pw_name, "ppuser"))
				{
				answer = 1;		/* Change to "true". */
				}
			}
		} /* end of if answer not determined yet */

	return answer ? TRUE : FALSE;
	} /* end of privledged() */

/*
** Return non-zero if the user is a PPR user charge
** account administrator.  Also, print an error message.
*/
static int am_administrator(void)
	{
	if( privledged() )
		{
		return TRUE;
		}
	else
		{
		fputs(_("You are not allowed to perform the requested operation because\n"
				"you are not a PPR user charge account administrator.\n"), stderr);
		return FALSE;
		}
	} /* end of am_administrator() */

/*
** Parse a float, first trying in the user's locale, and then,
** if that doesn't work, we try to parse a float in the notation
** of the C locale.
*/
static int parse_float(const char string[], float *store_here)
	{
	char *stop_ptr;

	*store_here = (float)strtod(string, (char**)&stop_ptr);

	if(!*stop_ptr && isspace(*stop_ptr))
		return 1;

	return gu_sscanf(string, "%f", store_here);
	}

/*===========================================================================
** Command action routines
===========================================================================*/

/*
** Add a record to the database.
*/
static int ppuser_add(char *argv[])
	{
	const char *username;
	struct userdb data;
	float x;

	if(! am_administrator() )
		return EXIT_DENIED;

	if(! argv[0] || ! argv[1] || ! argv[2] || ! argv[3] || ! argv[4] || ! argv[5])
		{
		fputs(_("Usage:\n"
			"\tppuser add <username> '<real name>' <authcode> <initial balance> <cutoff> <life>\n"), stderr);
		return EXIT_SYNTAX;
		}

	username = argv[0];

	strncpy(data.fullname, argv[1], MAX_FULLNAME);		/* the real name */
	data.fullname[MAX_FULLNAME] = '\0';

	strncpy(data.authcode, argv[2], MAX_AUTHCODE);		/* the AuthCode */
	data.authcode[MAX_AUTHCODE] = '\0';

	parse_float(argv[3], &x);					/* the initial balance */
	x *= 100.0;
	data.balance = (int)x;

	parse_float(argv[4], &x);					/* the credit cutoff point */
	x *= 100.0;
	data.cutoff = (int)x;

	gu_sscanf(argv[5], "%d", &data.lifetime);	/* how long to preserve */

	data.revoked = FALSE;						/* credit not revoked */

	switch(db_add_user(username, &data))
		{
		case USER_OK:
			log_ppuser("ppuser add \"%s\" \"%s\" %s %s %s %s\n",
				argv[0], argv[1], argv[2], argv[3], argv[4], argv[5]);
			return EXIT_OK;
		default:
			fprintf(stderr, _("Database error.\n"));
			return EXIT_INTERNAL;
		}
	} /* end of ppuser_add() */

/*
** Delete a record from the database.
*/
static int ppuser_delete(char *argv[])
	{
	const char *username;

	if(! am_administrator() )
		return EXIT_DENIED;

	if(! (username = argv[0]))
		{
		fprintf(stderr, _("Usage:\n\tppuser delete '<user name>'\n"));
		return EXIT_SYNTAX;
		}

	switch(db_delete_user(username))
		{
		case USER_ISNT:
			fprintf(stderr, _("The user \"%s\" does not exist.\n"), username);
			return EXIT_NOTFOUND;
		case USER_OK:
			log_ppuser("ppuser delete \"%s\"\n", username);
			return EXIT_OK;
		default:
			fprintf(stderr, _("Database error.\n"));
			return EXIT_INTERNAL;
		}
	} /* end of ppuser_delete() */

/*
** Show a user's recored.
*/
static int ppuser_show(char *argv[])
	{
	struct userdb entry;
	const char *username = argv[0];

	if(! username)
		{
		fprintf(stderr, _("Usage:\n\tppuser show '<user name>'\n"));
		return EXIT_SYNTAX;
		}

	switch(db_auth(&entry, username))
		{
		case USER_OK:
		case USER_OVERDRAWN:
			printf(_("Name: %s\n"), username);
			printf(_("Fullname: %s\n"), entry.fullname);
			printf(_("Balance: %s\n"), money(entry.balance));
			printf(_("Cutoff point: %s\n"), money(entry.cutoff));
			{
			struct tm *t;
			char temp[32];
			/* This one is for the i18n of the date format.  Should we follow 
			 * the format of the flag pages (as we do here), or should we use
			 * the format from the user's locale?
			 */
			char *format = gu_ini_query(PPR_CONF, "internationalization", "flagdateformat", 0, "%d-%b-%Y, %I:%M%p");
			t = localtime(&entry.last_mod);
			strftime(temp, sizeof(temp), "%d %b %Y %I:%M %p", t);
			strftime(temp, sizeof(temp), format, t);
			printf(_("Last Modified: %s\n"), temp);
			}
			printf(_("Account lifetime: %d\n"), entry.lifetime);
			printf(_("Credit revoked: %s\n"), entry.revoked ? _("TRUE") : _("FALSE"));
			return EXIT_OK;
		case USER_ISNT:
			printf(_("The user \"%s\" does not exist.\n"), username);
			return EXIT_NOTFOUND;
		default:
			fprintf(stderr, _("Database error.\n"));
			return EXIT_INTERNAL;
		}
	} /* end of ppuser_show() */

/*
** Deposit money in a users account.
*/
static int ppuser_transaction(char *argv[], enum TRANSACTION transaction_type)
	{
	const char *username;
	float x;
	int amount;
	int ret;

	if(! am_administrator() )
		return EXIT_DENIED;

	if(! (username = argv[0]) || ! argv[1])
		{
		fputs(_("Usage:\n"
			"\tppuser {deposit,withdraw,charge,correction} '<user name>' <amount>\n"), stderr);
		return EXIT_SYNTAX;
		}

	parse_float(argv[1], &x);
	x *= 100.0;
	amount = (int)x;

	ret = db_transaction(username, amount, transaction_type);

	switch(ret)
		{
		case USER_OK:
			switch(transaction_type)
				{
				case TRANSACTION_CHARGE:
					log_ppuser("ppuser charge \"%s\" %s\n", username, argv[1]);
					break;
				case TRANSACTION_DEPOSIT:
					log_ppuser("ppuser deposit \"%s\" %s\n", username, argv[1]);
					break;
				case TRANSACTION_WITHDRAWAL:
					log_ppuser("ppuser withdraw \"%s\" %s\n", username, argv[1]);
					break;
				case TRANSACTION_CORRECTION:
					log_ppuser("ppuser correction \"%s\" %s\n", username, argv[1]);
					break;
				}
			return EXIT_OK;
		case USER_ISNT:
			fprintf(stderr, _("The user \"%s\" does not exist.\n"), username);
			return EXIT_NOTFOUND;
		default:
			fprintf(stderr, _("Database error, transaction not processed.\n"));
			return EXIT_INTERNAL;
		}
	} /* end of ppuser_transaction() */

/*
** called from main() if no valid command
*/
static void help(FILE *outfile)
	{
	fputs(_("Usage:\n"
		"\tppuser add <username> '<real name>' <authcode> <initial balance> <cutoff> <life>\n"
		"\tppuser delete <username>\n"
		"\tppuser show <username>\n"
		"\tppuser deposit <username> amount\n"
		"\tppuser withdraw <username> amount\n"
		"\tppuser charge <username> amount\n"
		"\tppuser correction <username> amount\n"), outfile);
	} /* end of help */

/*
** main() function and command dispatcher
*/
int main(int argc, char *argv[])
	{
	/* Initialize internation messages library. */
	#ifdef INTERNATIONAL
	setlocale(LC_ALL, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);
	#endif

	/* parnoia */
	umask(PPR_UMASK);

	if(argc < 2)		/* if too few to be any valid command */
		{
		help(stderr);
		return EXIT_SYNTAX;
		}

	if(gu_strcasecmp(argv[1], "add") == 0)
		return ppuser_add(&argv[2]);
	else if(gu_strcasecmp(argv[1], "delete") == 0)
		return ppuser_delete(&argv[2]);
	else if(gu_strcasecmp(argv[1], "show") == 0)
		return ppuser_show(&argv[2]);
	else if(gu_strcasecmp(argv[1], "deposit") == 0)
		return ppuser_transaction(&argv[2],TRANSACTION_DEPOSIT);
	else if(gu_strcasecmp(argv[1], "withdraw") == 0)
		return ppuser_transaction(&argv[2],TRANSACTION_WITHDRAWAL);
	else if(gu_strcasecmp(argv[1], "charge") == 0)
		return ppuser_transaction(&argv[2],TRANSACTION_CHARGE);
	else if(gu_strcasecmp(argv[1], "correction") == 0)
		return ppuser_transaction(&argv[2],TRANSACTION_CORRECTION);
	else if(gu_strcasecmp(argv[1], "help") == 0)
		{
		help(stdout);
		return EXIT_OK;
		}
	else
		{
		help(stderr);
		return EXIT_SYNTAX;
		}
	} /* end of main() */

/* end of file */

